File: backend/routers/escalations.py
"""Escalations router — store and manage escalations to care providers."""

import uuid

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session, joinedload

from models.escalation import Escalation
from schemas.escalation import EscalationCreate, EscalationResponse, EscalationStatusUpdate
from services.database import get_db

router = APIRouter(prefix="/escalations", tags=["escalations"])

_VALID_URGENCY = {"low", "medium", "high"}
_VALID_STATUS = {"open", "acknowledged", "resolved"}


def _to_response(e: Escalation) -> EscalationResponse:
    return EscalationResponse.model_validate({
        "id": e.id,
        "patient_id": e.patient_id,
        "patient_name": f"{e.patient.first_name} {e.patient.last_name}",
        "session_id": e.session_id,
        "reason": e.reason,
        "urgency": e.urgency,
        "status": e.status,
        "notification_status": e.notification_status,
        "created_at": e.created_at,
    })


@router.post("/", response_model=EscalationResponse, status_code=201)
def create_escalation(body: EscalationCreate, db: Session = Depends(get_db)) -> EscalationResponse:
    """Store an escalation. Called by the MCP server tool escalate_to_human."""
    if body.urgency not in _VALID_URGENCY:
        raise HTTPException(status_code=422, detail=f"urgency moet een van {_VALID_URGENCY} zijn")

    escalation = Escalation(
        patient_id=body.patient_id,
        session_id=body.session_id,
        reason=body.reason,
        urgency=body.urgency,
        status="open",
        notification_status="pending",
    )
    db.add(escalation)
    db.commit()
    db.refresh(escalation)
    db.refresh(escalation, ["patient"])

    # Issue #25: send notification here (email for low/medium, Slack for high)
    # and update escalation.notification_status to "sent" or "failed".

    return _to_response(escalation)


@router.get("/", response_model=list[EscalationResponse])
def list_escalations(db: Session = Depends(get_db)) -> list[EscalationResponse]:
    """Return all escalations, newest first."""
    rows = (
        db.query(Escalation)
        .options(joinedload(Escalation.patient))
        .order_by(Escalation.created_at.desc())
        .all()
    )
    return [_to_response(e) for e in rows]


@router.get("/{escalation_id}", response_model=EscalationResponse)
def get_escalation(escalation_id: uuid.UUID, db: Session = Depends(get_db)) -> EscalationResponse:
    """Return one escalation by ID."""
    escalation = (
        db.query(Escalation)
        .options(joinedload(Escalation.patient))
        .filter(Escalation.id == escalation_id)
        .first()
    )
    if not escalation:
        raise HTTPException(status_code=404, detail="Escalatie niet gevonden")
    return _to_response(escalation)


@router.patch("/{escalation_id}/status", response_model=EscalationResponse)
def update_escalation_status(
    escalation_id: uuid.UUID,
    body: EscalationStatusUpdate,
    db: Session = Depends(get_db),
) -> EscalationResponse:
    """Update status (open → acknowledged → resolved)."""
    if body.status not in _VALID_STATUS:
        raise HTTPException(status_code=422, detail=f"status moet een van {_VALID_STATUS} zijn")

    escalation = (
        db.query(Escalation)
        .options(joinedload(Escalation.patient))
        .filter(Escalation.id == escalation_id)
        .first()
    )
    if not escalation:
        raise HTTPException(status_code=404, detail="Escalatie niet gevonden")

    escalation.status = body.status
    db.commit()
    db.refresh(escalation)
    return _to_response(escalation)

File: frontend/Anna-remembers/lib/api/escalations.ts
import { get, patch } from "./_base"
import type { Escalation, EscalationUrgency, EscalationStatus } from "@/types"

// ─── Backend response type ─────────────────────────────────────────

interface EscalationAPI {
  id: string
  patient_id: string
  patient_name: string
  session_id: string | null
  reason: string
  urgency: "low" | "medium" | "high"
  status: "open" | "acknowledged" | "resolved"
  notification_status: string
  notes: string | null
  created_at: string
}

// ─── Mapping ──────────────────────────────────────────────────────

const URGENCY_MAP: Record<EscalationAPI["urgency"], EscalationUrgency> = {
  high: "urgent",
  medium: "warning",
  low: "info",
}

const STATUS_MAP: Record<EscalationAPI["status"], EscalationStatus> = {
  open: "open",
  acknowledged: "in_progress",
  resolved: "closed",
}

const STATUS_MAP_REVERSE: Record<EscalationStatus, EscalationAPI["status"]> = {
  open: "open",
  in_progress: "acknowledged",
  closed: "resolved",
}

const CHANNEL_MAP: Record<EscalationAPI["urgency"], string> = {
  high: "Slack",
  medium: "E-mail",
  low: "E-mail",
}

function toEscalation(e: EscalationAPI): Escalation {
  return {
    id: e.id,
    patient: e.patient_id,
    name: e.patient_name,
    urgency: URGENCY_MAP[e.urgency],
    status: STATUS_MAP[e.status],
    reason: e.reason,
    channel: CHANNEL_MAP[e.urgency],
    assignee: null,
    opened: e.created_at,
    closed: null,
    notes: e.notes,
  }
}

// ─── Exports ──────────────────────────────────────────────────────

export async function getEscalations(): Promise<Escalation[]> {
  const data = await get<EscalationAPI[]>("/escalations/")
  return data.map(toEscalation)
}

export async function updateEscalationStatus(
  id: string,
  status: EscalationStatus,
  notes?: string
): Promise<Escalation> {
  const data = await patch<EscalationAPI>(`/escalations/${id}/status`, {
    status: STATUS_MAP_REVERSE[status],
    ...(notes !== undefined && { notes }),
  })
  return toEscalation(data)
}
